net/message.go

```diff
@@ -0,0 +1,95 @@
+package net
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/athanorlabs/go-relayer/common"
+)
+
+// Message must be implemented by all network messages
+type Message interface {
+	Type() byte
+}
+
+func decodeMessage(b []byte) (Message, error) {
+	// 1-byte type followed by at least 2-bytes of JSON (`{}`)
+	if len(b) < 3 {
+		return nil, errors.New("invalid message bytes")
+	}
+
+	msgType := b[0]
+	msgJSON := b[1:]
+	var msg Message
+
+	switch msgType {
+	case TransactionResponseType:
+		msg = &TransactionResponse{}
+	case TransactionRequestType:
+		msg = &TransactionRequest{}
+	default:
+		return nil, fmt.Errorf("invalid message type %d", msgType)
+	}
+
+	if err := json.Unmarshal(msgJSON, &msg); err != nil {
+		return nil, fmt.Errorf("failed to decode message with type %d: %w", msg.Type(), err)
+	}
+	return msg, nil
+}
+
+const (
+	TransactionRequestType byte = iota
+	TransactionResponseType
+)
+
+type TransactionRequest struct {
+	*common.SubmitTransactionRequest
+}
+
+// String ...
+func (m *TransactionRequest) String() string {
+	return fmt.Sprintf("TransactionRequest %v", m.SubmitTransactionRequest)
+}
+
+// Encode ...
+func (m *TransactionRequest) Encode() ([]byte, error) {
+	b, err := json.Marshal(m)
+	if err != nil {
+		return nil, err
+	}
+
+	return append([]byte{TransactionRequestType}, b...), nil
+}
+
+func (r *TransactionRequest) Type() byte {
+	return TransactionRequestType
+}
+
+type TransactionResponse struct {
+	*common.SubmitTransactionResponse
+	Error error
```

Comment:
Will there ever be more than one field in `SubmitTransactionResponse`? If not, you could just add the transaction ID here and copy it over.  If you leave it the way it is, maybe add a comment that accessing the TxHash will segfault if you don't first verify that the error value is nil.

Reply:
I'll make it a non pointer and add `omitempty` tags, and validate that only one is set. `SubmitTransaction` already returns an error if the error is set